morl_baselines/multi_policy/pareto_q_learning/pql.py

```diff
@@ -1,312 +1,244 @@
+            "epsilon_decay": self.epsilon_decay,
+            "final_epsilon": self.final_epsilon,
+            "seed": self.seed
+        }
+
+    def score_pareto_cardinality(self, state):
+        """Compute the action scores based upon the Pareto cardinality metric.
 
-        transition_probs = self.transitions[state, action] / max(1, np.sum(self.transitions[state, action]))
-        next_states = np.where(self.transitions[state, action, :] > 0)[0]  # Next states with prob > 0
+        Args:
+            state (int): The current state.
 
-        next_sets = []
-        for next_state in next_states:
-            next_sets.append(list(self.non_dominated[state][action][next_state]))
+        Returns:
+            ndarray: A score per action.
+        """
+        q_sets = [self.get_q_set(state, action) for action in range(self.num_actions)]
+        candidates = set().union(*q_sets)
+        non_dominated = get_non_dominated(candidates)
+        scores = np.zeros(self.num_actions)
 
-        cartesian_product = itertools.product(*next_sets)
+        for vec in non_dominated:
+            for action, q_set in enumerate(q_sets):
+                if vec in q_set:
+                    scores[action] += 1
 
-        for next_vectors in cartesian_product:
-            expected_vec = np.zeros(self.num_objectives)
-            for idx, next_vector in enumerate(next_vectors):
-                next_state = next_states[idx]
-                transition_prob = transition_probs[next_state]
-                disc_future_reward = self.gamma * np.array(next_vector)
-                expected_vec += transition_prob * (self.avg_r[state, action, next_state] + disc_future_reward)
-            expected_vec = tuple(np.around(expected_vec, decimals=self.decimals))  # Round the future reward.
-            q_set.add(tuple(expected_vec))
-        return q_set
+        return scores
 
-    def select_action(self, state):
-        """Select an action using epsilon greedy on the performance metric.
+    def score_hypervolume(self, state):
+        """Compute the action scores based upon the hypervolume metric.
 
         Args:
             state (int): The current state.
 
         Returns:
-            int: The selected action.
+            ndarray: A score per action.
         """
-        if np.random.uniform(0, 1) < self.epsilon:
-            return np.random.randint(self.num_actions)
-        else:
-            values = []
-            for action in range(self.num_actions):
-                q_set = self.calc_q_set(state, action)
-                q_set = get_non_dominated(q_set)
-                value = self.perf_indic(q_set)
-                values.append(value)
-            action = np.random.choice(np.argwhere(values == np.max(values)).flatten())
-            return action
+        q_sets = [self.get_q_set(state, action) for action in range(self.num_actions)]
+        action_scores = [self.hypervolume(-1 * np.array(list(q_set))) for q_set in q_sets]
```

Comment:
I guess you can use the hypervolume function in `common` which does the inversion for you.

Reply:
I thought about doing that, but due to the implementation in ```common```, that way every time you compute the hypervolume a new ```HV``` object is created. I think that will be much slower.

Reply:
Might be worth trying it. Or modify the common implementation to return the HV object ? :-)

Reply:
I now use the implementation from common